// playcanvas/build/playcanvas/src/scene/gsplat-unified/gsplat-placement.js
import { GSplatStreams } from '../gsplat/gsplat-streams.js';
import { WORKBUFFER_UPDATE_ALWAYS, WORKBUFFER_UPDATE_ONCE, WORKBUFFER_UPDATE_AUTO } from '../constants.js';

class GSplatPlacement {
		destroy() {
				this._streams?.destroy();
				this._streams = null;
				this.intervals.clear();
				this.resource = null;
		}
		set workBufferModifier(value) {
				this._workBufferModifier = value;
				this.renderDirty = true;
		}
		get workBufferModifier() {
				return this._parentPlacement?.workBufferModifier ?? this._workBufferModifier;
		}
		consumeRenderDirty() {
				const format = this.resource?.format;
				if (format && this._lastFormatVersion !== format.extraStreamsVersion) {
						this._lastFormatVersion = format.extraStreamsVersion;
						this.renderDirty = true;
				}
				if (this.workBufferUpdate === WORKBUFFER_UPDATE_ALWAYS) {
						this.renderDirty = true;
				} else if (this.workBufferUpdate === WORKBUFFER_UPDATE_ONCE) {
						this.renderDirty = true;
						this.workBufferUpdate = WORKBUFFER_UPDATE_AUTO;
				}
				const dirty = this.renderDirty;
				this.renderDirty = false;
				return dirty;
		}
		set aabb(aabb) {
				this._aabb = aabb?.clone() ?? null;
		}
		get aabb() {
				const aabb = this._aabb ?? this.resource?.aabb;
				return aabb;
		}
		set lodDistances(distances) {
				const isOctree = !!(this.resource && this.resource.octree);
				if (isOctree) {
						if (distances) {
								this.resource.octree?.lodLevels ?? 1;
								this._lodDistances = distances.slice();
						} else {
								this._lodDistances = null;
						}
				}
		}
		get lodDistances() {
				return this._lodDistances ? this._lodDistances.slice() : null;
		}
		getInstanceTexture(name, device) {
				const resource = this.resource;
				if (!resource?.format) {
						return undefined;
				}
				if (!this._streams && resource.format.instanceStreams.length > 0) {
						this._streams = new GSplatStreams(device, true);
						this._streams.textureDimensions.copy(resource.streams.textureDimensions);
						this._streams.syncWithFormat(resource.format);
				}
				return this._streams?.getTexture(name);
		}
		get streams() {
				return this._parentPlacement?.streams ?? this._streams;
		}
		ensureInstanceStreams(device) {
				const resource = this.resource;
				if (!resource?.format) {
						return;
				}
				if (!this._streams && resource.format.instanceStreams.length > 0) {
						this._streams = new GSplatStreams(device, true);
						this._streams.textureDimensions.copy(resource.streams.textureDimensions);
						this._streams.syncWithFormat(resource.format);
				}
		}
		constructor(resource, node, lodIndex = 0, parameters = null, parentPlacement = null, id = null){
				this.intervals = new Map();
				this.id = 0;
				this.lodIndex = 0;
				this._lodDistances = null;
				this.splatBudget = 0;
				this._aabb = null;
				this.parameters = null;
				this._streams = null;
				this.renderDirty = false;
				this.workBufferUpdate = WORKBUFFER_UPDATE_AUTO;
				this._lastFormatVersion = -1;
				this._workBufferModifier = null;
				this._parentPlacement = null;
				this.id = id ?? parentPlacement?.id ?? 0;
				this.resource = resource;
				this.node = node;
				this.lodIndex = lodIndex;
				this.parameters = parameters ?? parentPlacement?.parameters ?? null;
				this._parentPlacement = parentPlacement;
		}
}

export { GSplatPlacement };
