// playcanvas/build/playcanvas/src/scene/gsplat-unified/gsplat-params.js
import { PIXELFORMAT_R32U, PIXELFORMAT_RGBA16F, PIXELFORMAT_RGBA16U, PIXELFORMAT_RGBA32U, PIXELFORMAT_RG32U } from '../../platform/graphics/constants.js';
import { ShaderMaterial } from '../materials/shader-material.js';
import { GSplatFormat } from '../gsplat/gsplat-format.js';

class GSplatParams {
		set colorizeLod(value) {
				if (this._colorizeLod !== value) {
						this._colorizeLod = value;
						this.dirty = true;
				}
		}
		get colorizeLod() {
				return this._colorizeLod;
		}
		set enableIds(value) {
				if (value && !this._enableIds) {
						this._enableIds = true;
						if (!this._format.getStream('pcId')) {
								this._format.addExtraStreams([
										{
												name: 'pcId',
												format: PIXELFORMAT_R32U
										}
								]);
						}
						this.dirty = true;
				}
		}
		get enableIds() {
				return this._enableIds;
		}
		set lodBehindPenalty(value) {
				if (this._lodBehindPenalty !== value) {
						this._lodBehindPenalty = value;
						this.dirty = true;
				}
		}
		get lodBehindPenalty() {
				return this._lodBehindPenalty;
		}
		set lodRangeMin(value) {
				if (this._lodRangeMin !== value) {
						this._lodRangeMin = value;
						this.dirty = true;
				}
		}
		get lodRangeMin() {
				return this._lodRangeMin;
		}
		set lodRangeMax(value) {
				if (this._lodRangeMax !== value) {
						this._lodRangeMax = value;
						this.dirty = true;
				}
		}
		get lodRangeMax() {
				return this._lodRangeMax;
		}
		set lodUnderfillLimit(value) {
				if (this._lodUnderfillLimit !== value) {
						this._lodUnderfillLimit = value;
						this.dirty = true;
				}
		}
		get lodUnderfillLimit() {
				return this._lodUnderfillLimit;
		}
		set splatBudget(value) {}
		get splatBudget() {
				return 0;
		}
		set colorRamp(value) {
				if (this._colorRamp !== value) {
						this._colorRamp = value;
						this.dirty = true;
				}
		}
		get colorRamp() {
				return this._colorRamp;
		}
		get material() {
				return this._material;
		}
		get format() {
				return this._format;
		}
		frameEnd() {
				this._material.dirty = false;
				this.dirty = false;
		}
		constructor(device){
				this._material = new ShaderMaterial();
				this.debugAabbs = false;
				this.radialSorting = false;
				this.gpuSorting = false;
				this.debugNodeAabbs = false;
				this.dirty = false;
				this._colorizeLod = false;
				this._enableIds = false;
				this.lodUpdateDistance = 1;
				this.lodUpdateAngle = 0;
				this._lodBehindPenalty = 1;
				this._lodRangeMin = 0;
				this._lodRangeMax = 10;
				this._lodUnderfillLimit = 0;
				this._colorRamp = null;
				this.colorRampIntensity = 1;
				this.colorizeColorUpdate = false;
				this.colorUpdateDistance = 0.2;
				this.colorUpdateAngle = 2;
				this.colorUpdateDistanceLodScale = 2;
				this.colorUpdateAngleLodScale = 2;
				this.cooldownTicks = 100;
				const colorFormat = device.getRenderableHdrFormat([
						PIXELFORMAT_RGBA16F
				]) || PIXELFORMAT_RGBA16U;
				this._format = new GSplatFormat(device, [
						{
								name: 'dataColor',
								format: colorFormat
						},
						{
								name: 'dataTransformA',
								format: PIXELFORMAT_RGBA32U
						},
						{
								name: 'dataTransformB',
								format: PIXELFORMAT_RG32U
						}
				], {
						readGLSL: '#include "gsplatContainerPackedReadVS"',
						readWGSL: '#include "gsplatContainerPackedReadVS"'
				});
		}
}

export { GSplatParams };
